File: mobile/flutter/lib/features/auth/application/use_cases.dart
import 'package:stackbraid_client/stackbraid_client.dart';

import '../../../shared/auth/session_port.dart';
import '../data/auth_repository.dart';
import '../domain/validation.dart';

class ValidationFailed implements Exception {
  ValidationFailed(this.fieldErrors);

  final FieldErrors fieldErrors;

  @override
  String toString() => 'ValidationFailed($fieldErrors)';
}

class AuthUseCases {
  AuthUseCases(this._repository);

  final AuthRepository _repository;

  /// `POST /v1/auth/register` only creates the account — the contract is
  /// explicit that it does not issue a session (unlike login/refresh, it
  /// returns the created `User`, not a `TokenPair`). Signing in right after,
  /// with the same credentials this form already collected, is this app's
  /// choice of UX, not something the endpoint does on its own — the same
  /// choice both web frontends make.
  Future<User> registerAccount(RegisterFormValues values, SessionPort session) async {
    final fieldErrors = validateRegisterForm(values);
    if (fieldErrors.isNotEmpty) throw ValidationFailed(fieldErrors);

    await _repository.register(values);
    final tokens = await _repository.signIn(
      LoginFormValues(email: values.email, password: values.password),
    );
    return session.adoptSession(tokens);
  }

  Future<User> signIn(LoginFormValues values, SessionPort session) async {
    final fieldErrors = validateLoginForm(values);
    if (fieldErrors.isNotEmpty) throw ValidationFailed(fieldErrors);

    final tokens = await _repository.signIn(values);
    return session.adoptSession(tokens);
  }

  Future<void> signOut(String? refreshToken, SessionPort session) async {
    try {
      await _repository.signOut(refreshToken);
    } catch (_) {
      // Swallowed deliberately: the session is cleared locally either way,
      // so an offline (or already-expired) sign-out still signs this device
      // out rather than surfacing a network error for an action that, from
      // the user's point of view, already succeeded. The web frontends make
      // the same choice in `shared/auth`.
    } finally {
      await session.clearSession();
    }
  }
}

File: frontends/angular/src/web/features/auth/application/use-cases.ts
import type { User } from "@stackbraid/client-typescript";

import type { SessionPort } from "../../../../shared/auth/auth.service";
import { authRepository } from "../data/auth-repository";
import { type LoginFormValues, type RegisterFormValues, validateLoginForm, validateRegisterForm } from "../domain/validation";

export class ValidationFailed extends Error {
  constructor(readonly fieldErrors: Record<string, string>) {
    super("One or more fields need attention.");
    this.name = "ValidationFailed";
  }
}

export async function registerAccount(values: RegisterFormValues, session: SessionPort): Promise<User> {
  const fieldErrors = validateRegisterForm(values);
  if (Object.keys(fieldErrors).length > 0) throw new ValidationFailed(fieldErrors as Record<string, string>);

  // `POST /v1/auth/register` only creates the account — the contract is
  // explicit that it does not issue a session (unlike login/refresh, it
  // returns the created `User`, not a `TokenPair`). Signing in right after,
  // with the same credentials this form already collected, is this app's
  // choice of UX, not something the endpoint does for it.
  await authRepository.register(values);
  const tokens = await authRepository.signIn({ email: values.email, password: values.password });
  return session.adoptSession(tokens);
}

export async function signIn(values: LoginFormValues, session: SessionPort): Promise<User> {
  const fieldErrors = validateLoginForm(values);
  if (Object.keys(fieldErrors).length > 0) throw new ValidationFailed(fieldErrors as Record<string, string>);

  const tokens = await authRepository.signIn(values);
  return session.adoptSession(tokens);
}

export async function signOut(session: SessionPort): Promise<void> {
  try {
    await authRepository.signOut();
  } catch {
    // Swallowed deliberately: the session is cleared locally either way, so
    // an offline (or already-expired) logout still logs this tab out rather
    // than surfacing a network error for an action that, from the user's
    // point of view, already succeeded.
  } finally {
    session.clearSession();
  }
}